### PR TITLE
Add support for 'Attribute' option on KAIROSDB output format

### PR DESCRIPTION
**Pull request by jaroug**
Support attributes on write_http module specificly on kairosdb format.

### CODE REVIEW

**Comment by rubenk** [comment]
Could you keep the same commit messages as your previous PR? That explains better what this patch does.

File: src/utils_format_kairosdb.c [hunk 1]
```diff
@@ -181,7 +181,9 @@ static int values_to_kairosdb(char *buffer, size_t buffer_size, /* {{{ */
 
 static int value_list_to_kairosdb(char *buffer, size_t buffer_size, /* {{{ */
                                   const data_set_t *ds, const value_list_t *vl,
-                                  int store_rates) {
+                                  int store_rates,
+                                  char const *const *http_attrs,
+                                  const int http_attrs_num) {
```
Comment (rubenk) [inline src/utils_format_kairosdb.c]: I reviewed this before, but that probably got lost. `const int` makes no sense in function parameters, please make this just an int. Also mentioned before, you declare this in write_http.c as a size_t, so pick one or the other.

File: src/utils_format_kairosdb.h [hunk 2]
```diff
@@ -35,11 +35,14 @@
 #define JSON_GAUGE_FORMAT GAUGE_FORMAT
 #endif
 
+
```
Comment (rubenk) [inline src/utils_format_kairosdb.h]: remove this unnecessary change please.

File: src/write_http.c [hunk 3]
```diff
@@ -468,9 +471,9 @@ static int wh_write_kairosdb(const data_set_t *ds,
     }
   }
 
-  status = format_kairosdb_value_list(cb->send_buffer, &cb->send_buffer_fill,
-                                      &cb->send_buffer_free, ds, vl,
-                                      cb->store_rates);
+  status = format_kairosdb_value_list(
+      cb->send_buffer, &cb->send_buffer_fill, &cb->send_buffer_free, ds, vl,
+      cb->store_rates, (char const *const *)http_attrs, http_attrs_num);
```
Comment (rubenk) [inline src/write_http.c]: unnecessary cast

File: src/write_http.c [hunk 4]
```diff
@@ -479,9 +482,9 @@ static int wh_write_kairosdb(const data_set_t *ds,
       return (status);
     }
 
-    status = format_kairosdb_value_list(cb->send_buffer, &cb->send_buffer_fill,
-                                        &cb->send_buffer_free, ds, vl,
-                                        cb->store_rates);
+    status = format_kairosdb_value_list(
+        cb->send_buffer, &cb->send_buffer_fill, &cb->send_buffer_free, ds, vl,
+        cb->store_rates, (char const *const *)http_attrs, http_attrs_num);
```
Comment (rubenk) [inline src/write_http.c]: same here

File: src/write_http.c (hunk 3 above)
Comment (jaroug) [inline src/write_http.c]: won't compile without the casting

Comment (rubenk) [inline src/write_http.c]: I believe it should, what's the error?

File: src/write_http.c [hunk 5]
```diff
@@ -703,7 +706,34 @@ static int wh_config_node(oconfig_item_t *ci) /* {{{ */
       status = cf_util_get_boolean(child, &cb->log_http_error);
     else if (strcasecmp("Header", child->key) == 0)
       status = wh_config_append_string("Header", &cb->headers, child);
-    else {
+    else if (strcasecmp("Attribute", child->key) == 0) {
+      char *key = NULL;
+      char *val = NULL;
+
+      if (child->values_num != 2) {
+        WARNING("write_http plugins: Attribute need both a key and a value.");
```
Comment (rubenk) [inline src/write_http.c]: plugin, not plugins

File: src/write_http.c [hunk 6]
```diff
@@ -703,7 +706,34 @@ static int wh_config_node(oconfig_item_t *ci) /* {{{ */
...
+        break;
+      }
+      if (child->values[0].type != OCONFIG_TYPE_STRING ||
+          child->values[1].type != OCONFIG_TYPE_STRING) {
+        WARNING("write_http plugins: Attribute needs string arguments.");
```
Comment (rubenk) [inline src/write_http.c]: same here

File: src/utils_format_kairosdb.c (hunk 1 above)
Comment (jaroug) [inline src/utils_format_kairosdb.c]: ok, im switching it back to size_t then, strarray_add() take this type as second parameter

File: src/write_http.c (hunk 3 above)
Comment (jaroug) [inline src/write_http.c]: `src/write_http.c: In function ‘wh_write_kairosdb’:
src/write_http.c:477:24: error: passing argument 7 of ‘format_kairosdb_value_list’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       cb->store_rates, http_attrs, http_attrs_num);
                        ^~~~~~~~~~
In file included from src/write_http.c:31:0:
src/utils_format_kairosdb.h:40:5: note: expected ‘const char * const*’ but argument is of type ‘char **’
 int format_kairosdb_value_list(char *buffer, size_t *ret_buffer_fill,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors`

Comment (rubenk) [inline src/write_http.c]: you declare static char **http_attrs in write_http.c, so there's the mismatch. that should probably be `static const char **http_attrs;`

File: src/write_http.c [hunk 7]
```diff
@@ -703,7 +706,34 @@ static int wh_config_node(oconfig_item_t *ci) /* {{{ */
...
+        break;
+      }
+      if ((val = strdup(child->values[1].value.string)) == NULL) {
+        WARNING("cannot allocate memory for attribute value.");
+        sfree(key);
+        break;
+      }
+      strarray_add(&http_attrs, &http_attrs_num, key);
+      strarray_add(&http_attrs, &http_attrs_num, val);
+      DEBUG("write_http plugins: got attribute: %s => %s", key, val);
```
Comment (rubenk) [inline src/write_http.c]: plugin, not plugins.

File: src/write_http.c (hunk 3 above)
Comment (jaroug) [inline src/write_http.c]: If I change to `static const char **http_attrs` it trigger error on strarray_add() which take a char *** as first parameter : 

```
src/write_http.c: In function ‘wh_config_node’:
src/write_http.c:731:20: error: passing argument 1 of ‘strarray_add’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       strarray_add(&http_attrs, &http_attrs_num, key);
                    ^
In file included from src/write_http.c:28:0:
./src/daemon/common.h:380:5: note: expected ‘char ***’ but argument is of type ‘const char ***’
 int strarray_add(char ***ret_array, size_t *ret_array_len, char const *str);
     ^~~~~~~~~~~~
src/write_http.c:732:20: error: passing argument 1 of ‘strarray_add’ from incompatible pointer type [-Werror=incompatible-pointer-types]
       strarray_add(&http_attrs, &http_attrs_num, val);
                    ^
In file included from src/write_http.c:28:0:
./src/daemon/common.h:380:5: note: expected ‘char ***’ but argument is of type ‘const char ***’
 int strarray_add(char ***ret_array, size_t *ret_array_len, char const *str);
     ^~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Comment (rubenk) [inline src/write_http.c]: strarray_add should probably be changed then. I'll make a note do this, please ignore my comment for now :)

Comment (jaroug) [inline src/write_http.c]: alright, cool :) so what's next?

**Comment by dekonnection** [comment]
Need that feature :+1:
